third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/PingPong.cpp

```diff
@@ -17,41 +45,174 @@ namespace ttg = mlir::triton::gpu;
+class CriticalRegionManager {
+private:
+  /// Barrier ID range constants
+  /// This pass only uses named barriers 7 - 15 and reserves 0 - 6 for other
+  /// uses.
+  static constexpr unsigned MIN_BARRIER_ID = 7;
+  static constexpr unsigned MAX_BARRIER_ID = 15;
+
+public:
+  /// Current barrier ID to assign (range [MIN_BARRIER_ID, MAX_BARRIER_ID])
+  int barrierId;
+
+  /// Indicator of whether we run out of barrier IDs
+  bool barrierIdExhausted;
+
+  /// Map from pingpong region id to its barrier ID
+  llvm::DenseMap<int, std::pair<int, int>> pingpongIdToBarrierId;
+
+  /// Map from pingpong region id to its critical operations
+  llvm::DenseMap<int, SmallVector<Operation *>> pingpongIdToKeyOps;
 
-// Treat exp2, mulf, addf, reduce as expensive computation when data type is
-// a tensor type of 1D or higher.
-static bool isExpensiveComp(Operation *op) {
-  if (!isa<arith::MulFOp>(op) && !isa<math::Exp2Op>(op) &&
-      !isa<arith::AddFOp>(op) && !isa<mlir::triton::ReduceOp>(op))
+  /// Map from pingpong region id to operations that mark
+  /// the critical region's start and end
+  llvm::DenseMap<int, SmallVector<Operation *>> pingpongIdToPingBoundaryOps;
+  llvm::DenseMap<int, SmallVector<Operation *>> pingpongIdToPongBoundaryOps;
+
+  /// Map from pingpong region id to the participating thread number
+  llvm::DenseMap<int, int> pingpongIdToThreadNum;
+
+  CriticalRegionManager() {
+    // Initialize barrier ID to be MIN_BARRIER_ID
+    barrierId = MIN_BARRIER_ID;
+    barrierIdExhausted = false;
+  }
+
+  /// Check if an operation is registered as an expensive operation for the
+  /// given compute capability. Only considers ops with 2D+ shaped operands.
+  bool isExpensiveOp(Operation *op, int computeCapability) const {
+    llvm::StringRef opName = op->getName().getStringRef();
+
+    // Fast path for math.exp/math.exp2: check result rank directly
+    switch (computeCapability) {
+    case 90: // Hopper
+      // On Hopper, wgmma is expensive
+      if (isa<ttng::WarpGroupDotOp>(op)) {
+        // WarpGroupDotOp has its own verifier that checks the tensor shapes
+        // so we can directly put a WarpGroupDotOp into pingpong region
+        LDBG("Encounter a " << opName << " op on Hopper.");
+        return true;
+      }
+      break;
+    case 100: // Blackwell
+      // On Blackwell, exp/exp2 uses SFU which can be expensive for multi-dim
+      // tensors Blackwell increases performance for GEMM which is no longer a
+      // bottleneck
+      if (isa<math::ExpOp, math::Exp2Op>(op)) {
+        if (op->getNumResults() == 0)
+          return false;
+        LDBG("Encounter a " << opName << " op on Blackwell.");
+        Type resultType = op->getResult(0).getType();
+        int64_t rank = -1;
+        if (auto tensorTy = dyn_cast<RankedTensorType>(resultType)) {
+          rank = tensorTy.getRank();
+          LDBG("RankedTensorType, rank is " << rank);
+        } else if (auto shapedTy = dyn_cast<ShapedType>(resultType)) {
```

Comment:
When would a math operation produce a `ShapedType` that is not a `RankedTensorType`?

Reply:
math.exp/math.exp2 could operate on vector type (vector<...xf32>) which is a ShapedType but not a RankedTensorType.